### PR TITLE
Fix null handling in `LiteralConstraints`

### PR DESCRIPTION
**Pull request by ggevay**
This fixes the bug that Michael [found](https://github.com/MaterializeInc/materialize/pull/18666#issuecomment-1500746144) in `LiteralConstraints`. It also makes nullability inference smarter to clean the error from the output of Michael's [typechecker](https://github.com/MaterializeInc/materialize/pull/18666#issuecomment-1500746144). And some other minor things. (see below)

I think Surfaces was auto-tagged due to a minor change in EXPLAIN: Earlier we forgot to print the attributes of `IndexedFilter` nodes, and I added this now. (in the 2nd commit)

### Motivation

  * This PR fixes a recognized bug: https://github.com/MaterializeInc/materialize/issues/18674

  * This PR adds a feature that has not yet been specified:
    * Makes the nullability inference in `typ` smarter for `Constant`: Look through the constant set of rows, and see whether we actually have any nulls. (This is similar to what the key inference already (partially) does.)
    * I added non-nullability where we are creating the `Constant` in `LiteralConstraints`. This is not strictly necessary after the above change, but it doesn't take much, and it's good to be explicit.
    * Makes the nullability inference in `typ` smarter for `Join`, so that nullability as detectable by `typ` doesn't change during `LiteralConstraints`, to avoid the change always appearing in [Michael's type checker's](https://github.com/MaterializeInc/materialize/pull/18666#issuecomment-1500746144) output.
    * (I added a key where we are creating the `Constant` in `LiteralConstraints`. Unfortunately this is not currently being propagated through the `Join`, because our key inference for `Join` is not smart enough. (I think this won't cause an issue in the typechecker, because the key is not present also before performing `LiteralConstraints`.))

  * Adds printing of attributes for `IndexedFilter`s.

  * Moves `literal_constraints.slt` to the `transform` dir.

### Tips for reviewer

It's better to view the commits one-by-one. The first has all the important changes.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: test/sqllogictest/transform/literal_constraints.slt [hunk 1]
```diff
@@ -192,6 +192,27 @@ Used Indexes:
...
 EOF
 
+# `a = NULL` should NOT find the NULL in the table.
+
+query T multiline
+EXPLAIN WITH(arity, join_impls) SELECT * FROM t1
+WHERE a = NULL OR a = 2
+----
+Explained Query (fast path):
+  Filter (null OR (#0 = 2))
```
Comment (aalexandrov) [inline test/sqllogictest/transform/literal_constraints.slt:202]: How would the plan before the fix be different?

**Review by aalexandrov — APPROVED** [review]
LGTM!

**Comment by ggevay** [comment]
Merging this now.

@mgree, if you later find that `LiteralConstraints` is still showing up in the typechecker for some reason, then we'll fix it in a follow-up PR.

**Comment by mgree** [comment]
> @mgree, if you later find that `LiteralConstraints` is still showing up in the typechecker for some reason, then we'll fix it in a follow-up PR.

Sounds good! Sorry to miss this during the week.

**Pull request merged**